File: code/ch06-05/non_fp.erl
%% @author J D Eisenberg <jdavid.eisenberg@gmail.com>
%% @doc Generate a random set of teeth, with a certain
%% percentage expected to be bad.
%% @copyright 2013 J D Eisenberg
%% @version 0.1

-module(non_fp).
-export([generate_teeth/2, test_teeth/0]).

%% @doc Generate a list of lists, six numbers per tooth, giving random
%% pocket depths. Takes a string where T="there's a tooth there"
%% and F="no tooth"), and a float giving probability that a tooth is good.

-spec(generate_teeth(string(), float()) -> list(list(integer()))).

generate_teeth(TeethPresent, ProbGood) ->
  random:seed(now()),
  generate_teeth(TeethPresent, ProbGood, []).

%% @doc Helper function that adds tooth data to the ultimate result.

-spec(generate_teeth(string(), float(), [[integer()]]) -> [[integer()]]).

generate_teeth([], _Prob, Result) -> lists:reverse(Result);

generate_teeth([$F | Tail], ProbGood, Result) ->
  generate_teeth(Tail, ProbGood, [[0] | Result]);
  
generate_teeth([$T | Tail], ProbGood, Result) ->
  generate_teeth(Tail, ProbGood,
  [generate_tooth(ProbGood) | Result]).

-spec(generate_tooth(float()) -> list(integer())).

%% @doc Generates a list of six numbers for a single tooth. Choose a
%% random number between 0 and 1. If that number is less than the probability
%% of a good tooth, it sets the "base depth" to 2, otherwise it sets the base
%% depth to 3.

generate_tooth(ProbGood) ->
  Good = random:uniform() < ProbGood,
  case Good of
    true -> BaseDepth = 2;
    false -> BaseDepth = 3
  end,
  generate_tooth(BaseDepth, 6, []).

%% @doc Take the base depth, add a number in range -1..1 to it,
%% and add it to the list.

generate_tooth(_Base, 0, Result) -> Result;

generate_tooth(Base, N, Result) ->
  [Base + random:uniform(3) - 2 | generate_tooth(Base, N - 1, Result)].
  
test_teeth() ->
  TList = "FTTTTTTTTTTTTTTFTTTTTTTTTTTTTTTT",
  N = generate_teeth(TList, 0.75),
  print_tooth(N).

print_tooth([]) -> io:format("Finished.~n");
print_tooth([H|T]) ->
  io:format("~p~n", [H]),
  print_tooth(T).
  
  

File: code/ch06-04/non_fp.ex
defmodule NonFP do
  @moduledoc """ 
  Use non-functional-programming constructs to create
  a list of lists with random entries.
  
  from *Études for Elixir*, O'Reilly Media, Inc., 2013.
  Copyright 2013 by J. David Eisenberg.
  """ 
  @vsn 0.1 

  @doc """
  Generate a list of list representing pocket depths for a set of teeth.
  The first argument is a character list consisting of T and F for teeth
  that are present or absent. The second argument is the probability that
  any given tooth will be good.
  """
  @spec generate_pockets(list, number) :: list(list)
  
  def generate_pockets(tooth_list, prob_good) do
    :random.seed(:erlang.timestamp())
    generate_pockets(tooth_list, prob_good, [])
  end
  
  # When the tooth list is empty, return the final list of lists
  # If a tooth is present, generate a set of six depths and add it
  # to the result; otherwise add a [0] to the result.
  
  defp generate_pockets([], _, result) do
    Enum.reverse(result)
  end
  
  defp generate_pockets([head | tail], prob_good, result) when head == ?T do
    tooth = generate_tooth(prob_good)
    generate_pockets(tail, prob_good, [tooth | result])
  end
  
  defp generate_pockets([_head | tail], _prob_good, result) do
    generate_pockets(tail, _prob_good, [[0] | result])
  end
  
  @doc """
  Generate a set of six pocket depths for a tooth, given a probability
  that a tooth is good.
  """
  @spec generate_tooth(number) :: list(number)
  
  def generate_tooth(prob_good) do
    r = :random.uniform()
    if (r < prob_good) do
      base_depth = 2
    else
      base_depth = 3
    end
    generate_tooth(base_depth, 6, [])
  end
  
  def generate_tooth(_base, 0, result) do
    result
  end
  
  def generate_tooth(base, n, result) do
    delta = :random.uniform(3) - 2  # result will be -1, 0, or 1
    generate_tooth(base, n - 1, [base + delta | result])
  end
  
  def test_pockets() do
    tlist = 'FTTTTTTTTTTTTTTFTTTTTTTTTTTTTTTT'
    big_list = generate_pockets(tlist, 0.75)
    print_pockets(big_list)
  end
  
  def print_pockets([]), do: IO.puts("Finished.")
  
  def print_pockets([head | tail]) do
    IO.puts(inspect(head))
    print_pockets(tail)
  end
end

